R-package/R/xgb.DMatrix.R

```diff
@@ -73,6 +82,40 @@ xgb.DMatrix <- function(data, info = list(), missing = NA, silent = FALSE, nthre
       missing,
       as.integer(NVL(nthread, -1))
     )
+  } else if (is.data.frame(data)) {
+    ctypes <- unlist(lapply(data, function(x) {
+      if (is.factor(x)) {
+        if (!enable_categorical) {
+          stop(
+            "When factor type is used, the parameter `enable_categorical`",
+            " must be set to TRUE."
+          )
+        }
+        "c"
+      } else if (is.integer(x)) {
+        "int"
+      } else if (is.logical(x)) {
+        "i"
```

Comment:
Question there: I see this is also being used in the pandas adapter.

In R, boolean (logical) types are represented as C `int`, where possible values are FALSE (zero), NA (-INT_MAX), and TRUE (everything else), while python's bool type has only True and False.

I see you mention in a comment later that these get converted to numeric type, but the C++ code still checks for integer/logical-typed columns.

What would happen with these missing values encoded as -INT_MAX if the columns are supplied in their original types?

Reply:
Thank you for the suggestion, I removed the `unlist` as suggested in https://github.com/dmlc/xgboost/pull/9828#discussion_r1420623833 .